### DISCUSSION TITLE
How to extend JWT interface @5.0.0-beta.2

### DISCUSSION THREAD

**Discussion by dayblox** [nextauthjs/next-auth#8945] [Help]
### Summary

I'm adding a role from the Google provider `profile`, through the `jwt` and `session` callbacks
I managed to extend the `User` and `Session` interfaces, but can't seem to extend the `JWT`.

```TypeScript
import NextAuth, { DefaultSession } from "next-auth"
import "next-auth/jwt"
import Google from "next-auth/providers/google"

declare module "next-auth" {
  interface User {
    role: string
  }

  interface Session {
    user: {
      role: string
    } & DefaultSession["user"]
  }
}

declare module "next-auth/jwt" {
  interface JWT {
    role: string
  }
}

export const {
  handlers: { GET, POST },
  auth,
} = NextAuth({
  providers: [
    Google({
      profile(profile) {
        return { id: profile.sub, role: "admin", ...profile }
      },
    }),
  ],
  callbacks: {
    jwt({ token, user }) {
      if (user) token.role = user.role
      return token
    },
    session({ session, token }) {
      if (session.user) session.user.role = token.role
      return session
    },
  },
})
```

`token.role` is sadly `unknown`

Thanks in advance for your kind help! ♥

### Additional information

next-auth: 5.0.0-beta.2
next: 13.5.6

**Comment by dayblox** [comment]
I realize that `next-auth/jwt` is deprecated. I have tried to override `JWT` other ways but didn't manage to do so :/

**Comment by dayblox** [comment]
It turns out you have to use the modules `@auth/core/types` and `@auth/core/X` for the `X` submodule, `@auth/core/JWT` in this case.

**Reply by dayblox** [reply]
Would be nice to have this clearly stated in the docs.

**Comment by dayblox** [comment]
When switching to `pnpm`, this doesn't seem to work anymore since `pnpm` doesn't install `@auth/core` directly into `node_modules` but rather hardlinks it into the `.pnpm/` folder, any clue how to work around this?

**Reply by prateek041** [reply]
did you find a solution for this?

**Comment by greenhat616** [comment]
Maybe it is a bug introduced since `beta`. When I used `experimental` version, I could extend JWT via module `@auth/core/jwt`, and now it is broken.

**Comment by SuperEgo-Health** [comment]
On Nextjs 14, I can't get the JWT additional fields addition via the callback to work. The session object returned by `auth` won't have the new fields attached. Anyone else encountering the same error?

**Reply by AmandineTournay** [reply]
I have the same problem with NextJS 14 and next-auth@5.0.0-beta.4 unfortunately. The JWT type remains the same even when trying with the documentation code: [Documentation](https://authjs.dev/getting-started/typescript#module-augmentation)

_And it works for the module @auth/core/types for User and Session interface_

**Reply by TobiasFG** [reply]
@SuperEgo-Health & @AmandineTournay The following worked for me, not sure whether the type of provider can affect it though.

``` Typescript
declare module '@auth/core/types' {
	interface User {
		id: string
		custom1: string
		custom2: string
	}

	interface Session {
		user: {
			id: string
			custom1: string
			custom2: string
		} & DefaultSession['user']
	}
}

declare module '@auth/core/jwt' {
	interface JWT {
		id: string
		custom1: string
		custom2: string
	}
}

export const {
	handlers: { GET, POST },
	auth,
} = NextAuth({
	session: { strategy: 'jwt' },
	providers: [
		AzureAd({
			clientId: process.env.SECRET_AZURE_AD_CLIENT_ID as string,
			clientSecret: process.env.SECRET_AZURE_AD_CLIENT_SECRET as string,
			tenantId: process.env.SECRET_AZURE_AD_TENANT_ID as string,
			profile(profile: any, tokens) {
				const customProfile = {
					...profile,
					id: profile.oid,
					custom1: profile.custom1 ? profile.custom1 : "Custom1",
					custom2: profile.custom2 ? profile.custom2 : "Custom2",
				}
				return customProfile as typeof customProfile
			},
		}),
	],
	callbacks: {
		jwt({ token, trigger, session, user }) {
			if (trigger === 'update' && session) {
				return { ...token, ...session?.user }
			}

			return { ...token, ...user }
		},
		session({ session, token, user }) {
			session.user = token
			return session
		},
	},
})
```

**Reply by AmandineTournay** [reply]
That's not really beautiful to put it in the same file indeed, but that worked !
I kept the `decalre module '@auth/core/types'` inside the file `next-auth.d.ts`, and I moved the `declare module '@auth/core/jwt'` inside my auth configuration file.

Thank you @TobiasFG

**Reply by TobiasFG** [reply]
> That's not really beautiful to put it in the same file indeed, but that worked !
> I kept the `decalre module '@auth/core/types'` inside the file `next-auth.d.ts`, and I moved the `declare module '@auth/core/jwt'` inside my auth configuration file.

Was in a hurry and quickly grapped relevant pieces and pasted in together, but glad to hear you figured out a separation meaningful to you :-)
